lobbyboy/config.py

```diff
@@ -1,10 +1,153 @@
+import json
 import logging
+import time
+from collections import OrderedDict
+from dataclasses import dataclass, replace, field, fields, asdict
+from json import JSONDecodeError
+from pathlib import Path
+from typing import Dict, List, Tuple, OrderedDict as typeOrderedDict
+
 import toml
 
+from lobbyboy.utils import lb_dict_factory
+
 logger = logging.getLogger(__name__)
 
 
-def load_config(config_path):
-    config = toml.load(config_path)
-    logger.info("loading configs... config: {}".format(config))
-    return config
+@dataclass
+class LBConfigUser:
+    authorized_keys: str = ""
+    password: bool = False
+
+    def auth_key_pairs(self) -> List[Tuple]:
+        """
+
+        Returns:
+            tuple: (key_type, key_data)
+        """
+        return [tuple(ssh_key.split(maxsplit=1)) for ssh_key in self.authorized_keys.split("\n") if ssh_key]
+
+
+@dataclass
+class LBConfigProvider:
+    load_module: str
+    min_life_to_live: str = "50m"  # 最少存活时间
+    bill_time_unit: str = "55m"
+    private_key: str = "auto"
+    extra_ssh_keys: List[str] = field(default_factory=list)
+    favorite_droplets: List[str] = field(default_factory=list)
+    api_token: str = ""
+    destroy_safe_time: str = "3m"
+    server_name_prefix: str = "lobbyboy"
+
+    # todo unique configuration of each provider
+    vagrantfile: str = ""
+
+
+@dataclass
+class LBServerMeta:
+    provider_name: str
+    workspace: Path
+    server_name: str
+    server_host: str = "127.0.0.1"
+    server_user: str = "root"
+    server_port: int = 22
+    created_timestamp: int = field(default_factory=lambda: int(time.time()))
+    # TODO support below features when create server
+    # extra ssh args when connect to this server by ssh, eg: ["-o", "ProxyCommand=$jumpServer"]
+    ssh_extra_args: List[str] = field(default_factory=list)
+    # indicate whether this server is managed by lobbyboy or not.
+    manage: bool = True
+
+    def __post_init__(self):
+        if isinstance(self.workspace, str):
+            self.workspace = Path(self.workspace)
+
+    @property
+    def live_sec(self) -> int:
+        return int(time.time()) - self.created_timestamp
+
+
+@dataclass
+class LBConfig:
+    _file: Path
+    _raw: Dict = field(default_factory=dict)
+    data_dir: Path = None
+    user: Dict[str, LBConfigUser] = field(default_factory=dict)
+    provider: Dict[str, LBConfigProvider] = field(default_factory=dict)
+    listen_port: int = 12200
+    listen_ip: str = "0.0.0.0"
+    log_level: str = "DEBUG"
+    min_destroy_interval: str = "5m"
+    servers_file: str = "available_servers_db.json"
+
+    def __post_init__(self):
+        # TODO
+        if self.data_dir:
+            self.data_dir = Path(self.data_dir)
+        if any(1 for i in self.user.values() if not isinstance(i, LBConfigUser)):
```

Comment:
for `LBConfig`, we don't know how others initialize it, is it given the value of the type it needs or just a python dict, but when we use it, we need it's type is clear, so i check it on the `__post_init__`, it's ugly, but it's works for now, i'll fix it after.
as for reload, we can it's `reload` method, it will read the config file again and give a new `LBConfig` instance.

Reply:
will there be issues if we run the type convert every time without `if`?

Reply:
yes, we can't init dataclass use another dataclass, most of time can't get result we want, and i've update [this](https://github.com/laixintao/lobbyboy/pull/16/files#diff-257d494626c98f775cf556aee0bb2b7080375ef155d73de7cf4ad008d2b3c8c1R105-R109)